### PR TITLE
8304885: Reuse stale data to improve DNS resolver resiliency

### PR DESCRIPTION
**Pull request by mrserb**
I would like to get preliminary feedback about the provided patch.

Discussion on net-dev@ https://mail.openjdk.org/pipermail/net-dev/2023-March/020682.html

One of the main issue I try to solve is how the cache handle the intermittent DNS server outages due to overloading or network connection.

At the moment this cache can be configured by the application using the following two properties:
   (1) "networkaddress.cache.ttl"(30 sec) - cache policy for successful lookups
   (2) "networkaddress.cache.negative.ttl"(10 sec) - cache policy for negative lookups

The default timeout for positive responses is good enough to "have recent dns-records" and to "minimize the number of requests to the DNS server".

But the cache for the negative responses is problematic. This is a problem I would like to solve. Caching the negative response means that for **10** seconds the application will not be able to connect to the server.

Possible solutions:
  1. Decreasing timeout "for the negative responses": unfortunately more requests to the server at the moment of "DNS-outage" cause even more issues, since this is not the right moment to load the network/server more.
  2. Increasing timeout "for the positive responses": this will decrease the chance to get an error, but the cache will start to use stale data longer.
  3. This proposal: it would be good to ignore the negative response and continue to use the result of the last "successful lookup" until some additional timeout.

The idea is to split the notion of the TTL and the timeout used for the cache. When TTL for the record will expire we should request the new data from the server. If this request goes fine we will update the record, if it fails we will continue to use the cached date until the next sync.

For example, if the new property "networkaddress.cache.extended.ttl" is set to 10 minutes, then we will cache a positive response for 10 minutes but will try to sync it every 30 seconds. If the new property is not set then as before we will cache positive for 30 seconds and then cache the negative response for 10 seconds.

RFC 8767 Serving Stale Data to Improve DNS Resiliency:
https://www.rfc-editor.org/rfc/rfc8767

Comments about current and other possible implementations:
 * The code intentionally moved to the separate ValidAddresses class, just to make clear that the default configuration, when the new property is not set is not changed much.
 * The refresh timeout includes the time spent on the server lookup. So if we have to refresh every 2 seconds, but in lookup, we spend 3 seconds then we will request the server on each lookup. Another implementation may spend 3 seconds on lookup and then additional use the cached value for two seconds.
 * The extended timeout is a kind of "maximum stale timer" from the RFC above, but it starts counting not from the moment the record expired, but from the moment we added it to the cache. Another possible implementation may start counting from the moment the TTL expired, so the overall usage of the value will be sum ttl+extended.
 * The extended timeout has a hard deadline which is never changed during execution, for example, if it sets for 1 day, then at the end of the day we should make a successful lookup to recache the value otherwise we will start to use the "negative" cache. Other implementations may shift the expiration time on every successful sync.

Any thoughts about other possibilities?

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8306653](https://bugs.openjdk.org/browse/JDK-8306653) to be approved

### Issues
 * [JDK-8304885](https://bugs.openjdk.org/browse/JDK-8304885): Reuse stale data to improve DNS resolver resiliency (**Enhancement** - P3)
 * [JDK-8306653](https://bugs.openjdk.org/browse/JDK-8306653): Reuse stale data to improve DNS resolver resiliency (**CSR**)

### Reviewers
 * [Michael McMahon](https://openjdk.org/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**)
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13285/head:pull/13285` \
`$ git checkout pull/13285`

Update a local copy of the PR: \
`$ git checkout pull/13285` \
`$ git pull https://git.openjdk.org/jdk.git pull/13285/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13285`

View PR using the GUI difftool: \
`$ git pr show -t 13285`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13285.diff">https://git.openjdk.org/jdk/pull/13285.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13285#issuecomment-1494130481)

### CODE REVIEW

**Comment by mrserb** [comment]
After thinking about the initial proposal and description of how it should work according to RFC 8767, I decided to change the meaning of the new property. Instead of setting the overall cache timeout via extended property, the code now uses the  "maximum stale timer" -  It defines the length of time after a record expires that it should be retained in the cache. It means that the overall timeout now is s sum of ttl+stale.

Note that now the "stale timeout" is updated after each successful lookup.

File: src/java.base/share/classes/java/net/InetAddress.java [hunk 1]
```diff
@@ -967,6 +972,99 @@ public int compareTo(CachedAddresses other) {
             // ties are broken using unique id
             return Long.compare(this.id, other.id);
         }
+
+        /**
+         * Checks if the current cache record is expired or not. Expired records
+         * are removed from the expirySet and cache.
+         */
+        public boolean expired(long now) {
```
Comment (djelinski) [inline src/java.base/share/classes/java/net/InetAddress.java:980]: This method's name does not reflect what this method does. It may be worth splitting.

File: src/java.base/share/classes/java/net/InetAddress.java [hunk 2]
```diff
@@ -967,6 +972,99 @@ public int compareTo(CachedAddresses other) {
...
+        {
+            super(host, inetAddresses, staleTime);
+            this.refreshTime = refreshTime;
+            this.staleTime = staleTime;
+        }
+
+        @Override
+        public InetAddress[] get() {
+            long now = System.nanoTime();
+            if ((refreshTime - now) < 0L && lookupLock.tryLock()) {
```
Comment (djelinski) [inline src/java.base/share/classes/java/net/InetAddress.java:1026]: I like it how you only block one thread while refreshing and let other threads use the stale value.

File: src/java.base/share/classes/sun/net/InetAddressCachePolicy.java [hunk 3]
```diff
@@ -121,40 +112,61 @@ public Integer run() {
...
 
         if (tmp != null) {
             negativeCachePolicy = tmp < 0 ? FOREVER : tmp;
             propertyNegativeSet = true;
         }
+        long max = TimeUnit.DAYS.toSeconds(7);
+        if (cachePolicy > 0 && cachePolicy < max) {
+            tmp = getProperty(cacheStalePolicyProp,
+                              cacheStalePolicyPropFallback);
+            if (tmp != null && tmp > cachePolicy) {
```
Comment (djelinski) [inline src/java.base/share/classes/sun/net/InetAddressCachePolicy.java:126]: Why do you require cacheStalePolicy > cachePolicy? In your implementation, stale timer starts running after the base timer expires, so lower values still make sense.

**Review by djelinski — COMMENTED** [review]
This looks like a nice improvement. You'll need to update the docs in all places where `networkaddress.cache.ttl` is mentioned, that is:
- https://github.com/openjdk/jdk/blob/0deb648985b018653ccdaf193dc13b3cf21c088a/src/java.base/share/classes/java/net/doc-files/net-properties.html#L263
- https://github.com/openjdk/jdk/blob/c6bd489cc8d30fb6eec865b3dab1cf861e25c8d7/src/java.base/share/classes/java/net/InetAddress.java#L200
- https://github.com/openjdk/jdk/blob/0deb648985b018653ccdaf193dc13b3cf21c088a/src/java.base/share/conf/security/java.security#L358

Plus, as Alan suggested, a CSR and a release note will also be necessary.

File: src/java.base/share/classes/java/net/InetAddress.java (hunk 1 above)
Comment (mrserb) [reply src/java.base/share/classes/java/net/InetAddress.java]: It is renamed to describe what it tries to do.

**Comment by mrserb** [comment]
> Plus, as Alan suggested, a CSR and a release note will also be necessary.

Sure, I just would like to get some initial feedback and have an agreement about the implementation/spec before submitting the CSR.

**Comment by mrserb** [comment]
The comments were updated as requested. I have also merged the master to make sure the GitHub actions will be green.

File: src/java.base/share/classes/sun/net/InetAddressCachePolicy.java [hunk 4]
```diff
@@ -121,40 +113,61 @@ public Integer run() {
...
             propertyNegativeSet = true;
         }
+        long max = TimeUnit.DAYS.toSeconds(7);
+        if (cachePolicy > 0 && cachePolicy < max) {
+            tmp = getProperty(cacheStalePolicyProp,
+                              cacheStalePolicyPropFallback);
+            if (tmp != null) {
+                staleCachePolicy = (int) Math.max(tmp, max);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (mrserb) [reply src/java.base/share/classes/sun/net/InetAddressCachePolicy.java:130]: It is come from the recommendation in the "RFC 8767" see the notion about "cap of 7 days":
https://www.rfc-editor.org/rfc/rfc8767
Depending on the final implementation I can document this, or delete it in the code.

[the remark this answers is not in the document]
Comment (Michael-Mc-Mahon) [reply src/java.base/share/classes/sun/net/InetAddressCachePolicy.java:130]: From what I can see RFC8767 recommends a cap of 7 days for the TTL itself - that is a maximum value for `cachePolicy` rather than `staleCachePolicy`. It seems to say the the stale timer should be configurable with the text: "The suggested value is between 1 and 3 days".  Personally, I would drop this part of the change. Such suggestions can be adopted (or not) by end users in my opinion.

[the remark this answers is not in the document]
Comment (mrserb) [reply src/java.base/share/classes/sun/net/InetAddressCachePolicy.java:130]: Sounds good, I'll update the patch.

**Review by Michael-Mc-Mahon — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/java/net/doc-files/net-properties.html [hunk 5]
```diff
@@ -267,13 +267,20 @@ <H2>Address Cache</H2>
 	policy, while a value of 0 (zero) means no caching. The default value
 	is -1 (forever) if a security manager is installed, and implementation-specific
 	when no security manager is installed.</P>
+	<LI><P><B>{@systemProperty networkaddress.cache.stale.ttl}</B> (default: see below)<BR>
+	Value is an integer corresponding to the number of seconds the stale names
+	will be kept in the cache. The name is considered stale if it expired and
+	attempts to refresh it were unsuccessful. This property is useful if it is
+	preferable to use a stale name rather than result of unsuccessful name lookup.
+	A value of 0 (zero) means do not use stale names.
+	The default value is implementation-specific.</P>
```
Comment (Michael-Mc-Mahon) [inline src/java.base/share/classes/java/net/doc-files/net-properties.html:276]:
```suggestion
	A value of 0 (zero) or if the property is not set means do not use stale names.
	By default, the property is not set.</P>
```

Comment (Michael-Mc-Mahon) [reply src/java.base/share/classes/java/net/doc-files/net-properties.html:276]: We probably should specify what happens if the stale property is negative. Looks like it might be ignored, which I guess is okay.

**Review by Michael-Mc-Mahon — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/java/net/doc-files/net-properties.html (hunk 5 above)
Comment (mrserb) [reply src/java.base/share/classes/java/net/doc-files/net-properties.html:276]: I have update the text about "property is not set" case, but leave the text about "default value" as implementation-specific, similarly to what we have in the text about "networkaddress.cache.ttl" property, and this allow to change this value in some situations.

Comment (mrserb) [reply src/java.base/share/classes/java/net/doc-files/net-properties.html:276]: Right now the negative value is ignored, and is not mentioned in the new spec, But it could be used to cover "always use the stale data on error" case. So we can change this in the future, but probably it will not be a big deal to update the spec if that will happen.

**Comment by mrserb** [comment]
If there are no any other comments/objections I'll start to work in the CSR.

**Comment by mrserb** [comment]
The CSR is updated, please take a look.

File: src/java.base/share/classes/java/net/doc-files/net-properties.html (hunk 5 above)
Comment (dfuch) [reply src/java.base/share/classes/java/net/doc-files/net-properties.html:276]: I believe we do need to specify what happens if the value is negative. We cannot leave that as "undefined behaviour that may change in the future".

File: src/java.base/share/classes/java/net/InetAddress.java [hunk 6]
```diff
@@ -933,23 +951,30 @@ public String toString() {
...
         InetAddress[] get() throws UnknownHostException;
     }
 
-    // a holder for cached addresses with required metadata
-    private static final class CachedAddresses  implements Addresses, Comparable<CachedAddresses> {
+    /**
+     * A cached result of a name service lookup. The result can be either valid
+     * addresses or invalid (ie a failed lookup) containing no addresses.
+     */
+    private static class CachedLookup implements Addresses, Comparable<CachedLookup> {
```
Comment (dfuch) [inline src/java.base/share/classes/java/net/InetAddress.java:966]: I do not really like the fact that CachedAddress -> CachedLookup has transformed an immutable class into a mutable one, which in addition forces a volatile read to get the addresses.

Have you explored moving the new functionality implemented by `ValidCachedLookup` inside `NameServiceAddresses`?

Couldn't NameServiceAddresses be initialized with a (final) InetAddress[] array (from the previous stale lookup) that, if not null, would be returned when `tryLock` returns false?

In other words if NameServiceAddresses::staleAddresses is not null, and we are still in the stale window, use tryLock, and if false, return staleAddresses; otherwise, use lock.

Would that work, and keep everything final instead of volatile?

Comment (mrserb) [reply src/java.base/share/classes/java/net/InetAddress.java:966]: This is what I tried to implement initially: NameServiceAddresses which stores the fallback value of staleAddresses. Unfortunately the NameServiceAddresses class is [used](https://github.com/openjdk/jdk/blob/aad899b7f964c4c4f05ad35208c58118692c9aa0/src/java.base/share/classes/java/net/InetAddress.java#L1808) when we already deleted the addresses from the caches. So the first thread may save the old stale data from the cache, delete and create a new "proper" NameServiceAddresses, but another thread may do the same and get the "empty" old stale data -> if the second request will be faster we will lost the stale data. It could be fixed by some additional synchronization, but that would be more expensive than one volatile read.

Note that we can eliminate that one volatile read by duplicating the logic of CachedLookup inside  ValidCachedLookup by using one interface for both, not sure that it is worth, since we already have a couple of loookups to the ConcurrentHashMap and ConcurrentSkipListSet on each call.

File: src/java.base/share/classes/java/net/InetAddress.java [hunk 7]
```diff
@@ -205,6 +206,23 @@
  * <p>
  * A value of -1 indicates "cache forever".
  * </dd>
+ * <dt><b>networkaddress.cache.stale.ttl</b></dt>
+ * <dd>Indicates the caching policy for stale names. The value is specified as
+ * an integer to indicate the number of seconds the stale names will be kept in
+ * the cache. The name is considered stale if it expired and attempts to refresh
+ * it were unsuccessful. This property is useful if it is preferable to use a
+ * stale name rather than result of unsuccessful name lookup. The default
+ * setting is to cache for an implementation specific period of time.
```
[the remark this answers is not in the document]
Comment (mrserb) [reply src/java.base/share/classes/java/net/InetAddress.java:215]: It is only true if the security property is not set for some specific binaries, and it can have some different value than zero by default.

File: src/java.base/share/classes/java/net/doc-files/net-properties.html (hunk 5 above)
Comment (mrserb) [reply src/java.base/share/classes/java/net/doc-files/net-properties.html:276]: will update the spec.

**Comment by mrserb** [comment]
@Michael-Mc-Mahon @djelinski @dfuch do you have any other comments? The CSR was updated based on feedback, please take a look.

File: src/java.base/share/classes/java/net/InetAddress.java [hunk 8]
```diff
@@ -205,6 +206,24 @@
...
+ * <p>
+ * If the value of this property is larger than "networkaddress.cache.ttl" then
+ * "networkaddress.cache.ttl" will be used as a refresh interval of the name in
+ * the cache. For example, if this property is set to 1 day and
+ * "networkaddress.cache.ttl" is set to 30 seconds, then the positive response
+ * will be cached for 1 day but an attempt to refresh it will be done every
+ * 30 seconds.
+ * <p>
+ * A value of 0 (zero) means do not use stale names, the negative value is
```
Comment (djelinski) [inline src/java.base/share/classes/java/net/InetAddress.java:224]: Should we be more specific here? As far as I could tell, negative values also disable caching.

Comment (mrserb) [reply src/java.base/share/classes/java/net/InetAddress.java]: I guess it is an accurate description since the negative value is ignored by the implementation.

**Comment by mrserb** [comment]
Since the RDP 1 for JDK 21 is coming, I would like to accelerate the review of this PR, do I need to provide some additional information to make it easy to review?

File: src/java.base/share/classes/java/net/doc-files/net-properties.html [hunk 9]
```diff
@@ -267,13 +267,21 @@ <H2>Address Cache</H2>
 	policy, while a value of 0 (zero) means no caching. The default value
 	is -1 (forever) if a security manager is installed, and implementation-specific
 	when no security manager is installed.</P>
+	<LI><P><B>{@systemProperty networkaddress.cache.stale.ttl}</B> (default: see below)<BR>
+	Value is an integer corresponding to the number of seconds the stale names
+	will be kept in the cache. The name is considered stale if it expired and
+	attempts to refresh it were unsuccessful. This property is useful if it is
+	preferable to use a stale name rather than result of unsuccessful name lookup.
+	A value of 0 (zero) or if the property is not set means do not use stale
+	names, the negative value is ignored.
```
Comment (dfuch) [inline src/java.base/share/classes/java/net/doc-files/net-properties.html:276]: I'd suggest to rephrase:

```suggestion
	A value of 0 (zero) or if the property is not set means do not use stale
	names. Negative values are ignored.
```

File: src/java.base/share/conf/security/java.security [hunk 10]
```diff
@@ -357,6 +357,16 @@ ssl.TrustManagerFactory.algorithm=PKIX
 #
 #networkaddress.cache.ttl=-1
 
+#
+# The Java-level namelookup cache stale policy:
+#
+# any positive value: the number of seconds to use the stale names
+# zero: do not use stale names
+#
+# default value is never (NEVER).
```
Comment (dfuch) [inline src/java.base/share/conf/security/java.security:366]:
```suggestion
# any positive value: the number of seconds to use the stale names
# zero: do not use stale names
# negative values are ignored
#
# default value is 0 (NEVER).
```

File: src/java.base/share/classes/java/net/doc-files/net-properties.html [hunk 11]
```diff
@@ -267,13 +267,21 @@ <H2>Address Cache</H2>
...
+	The default value is implementation-specific.</P>
```
Comment (dfuch) [inline src/java.base/share/classes/java/net/doc-files/net-properties.html:277]: Maybe we should say that the default value is 0 since this is what is documented in `java.security`

File: src/java.base/share/classes/java/net/InetAddress.java [hunk 12]
```diff
@@ -205,6 +206,24 @@
...
+ * <p>
+ * If the value of this property is larger than "networkaddress.cache.ttl" then
+ * "networkaddress.cache.ttl" will be used as a refresh interval of the name in
+ * the cache. For example, if this property is set to 1 day and
+ * "networkaddress.cache.ttl" is set to 30 seconds, then the positive response
+ * will be cached for 1 day but an attempt to refresh it will be done every
+ * 30 seconds.
+ * <p>
+ * A value of 0 (zero) or if the property is not set means do not use stale
+ * names, the negative value is ignored.
```
Comment (dfuch) [inline src/java.base/share/classes/java/net/InetAddress.java:225]:
```suggestion
 * names. Negative value are ignored.
```

File: src/java.base/share/classes/java/net/doc-files/net-properties.html (hunk 11 above)
Comment (mrserb) [reply src/java.base/share/classes/java/net/doc-files/net-properties.html:277]: In the "java.security" file we documented the default value of the property in that file but it is commented out. Right now the new property specified in the same way as the old "networkaddress.cache.ttl", see [net-properties.html](https://github.com/openjdk/jdk/pull/13285/files/ce595f74f511553274c9da254d53b368abe44408#diff-21ffe1f9a665c0f70cc9888d2da937a2a77fadfdacc23d0bd59506595e668c5fR267) and [InetAddress.java](https://github.com/openjdk/jdk/pull/13285/files/a856cd36eba1efacbc0f95918ac6aa2712cab870#diff-55d4acc64110543896bf6c289d6247cded986edfe9ccce3a9bca88a197711d58R204)

Comment (mrserb) [reply src/java.base/share/classes/java/net/doc-files/net-properties.html:277]: In the [InetAddress.java](https://github.com/openjdk/jdk/pull/13285/files/a856cd36eba1efacbc0f95918ac6aa2712cab870#diff-55d4acc64110543896bf6c289d6247cded986edfe9ccce3a9bca88a197711d58R204)
```
 * <dt><b>networkaddress.cache.ttl</b></dt>
 * <dd>Indicates the caching policy for successful name lookups from
 * the name service. The value is specified as an integer to indicate
 * the number of seconds to cache the successful lookup. The default
 * setting is to cache for an implementation specific period of time.
```
In the [java.security](https://github.com/openjdk/jdk/pull/13285/files/ce595f74f511553274c9da254d53b368abe44408#diff-)
```
# default value is forever (FOREVER). For security reasons, this
# caching is made forever when a security manager is set. When a security
# manager is not set, the default behavior in this implementation
# is to cache for 30 seconds.
```

Comment (mrserb) [reply src/java.base/share/classes/java/net/doc-files/net-properties.html:277]: Such specification allows changing the default property in JDK bundle w/o changing the doc.

**Comment by mrserb** [comment]
Thank you for your feedback, I am looking into this possible issue.

**Comment by mrserb** [comment]
The test expected to be completed in 2 seconds(10 lookups on a different threads). probably it is too short for mach5, will increase the time window.

**Comment by mrserb** [comment]
The test was expected to use two "refresh" time frames, in the first frame the dns record is cached, in the second frame the dns record should be updated by one thread only, all others should use the stale data. Initially the time-frame was equal to 2 seconds, now it is increased to 7 seconds. But even if the system is too slow the test now will ignore timestamp checks outside of 12 seconds which is less than 14 second(two frames).

**Comment by mrserb** [comment]
Another possible problem is that the Thread.sleep used by the test may not be accurate enough, I will change that as well.

**Comment by mrserb** [comment]
@dfuch the test is updated, please take a look, it includes:

 - The fix for the slow systems when it was not enough time to complete the test.
 - The fix for the possible race in the SimpleResolverProviderImpl:
    1 Thread_1 decided to refresh the record, it takes the lock in the InetAddress.java, but do not call the ResolverProvider yet.
    2 Thread_2 decided to get a lock, fail to do that, take a stale data, and save the timestamp
    3 Thread_1 call the ResolverProvider and update the timestamp, then intentionally hangs(becouse of the blocker)
    4 Thread_2 decided to get a lock, fail to do that again, take a stale data, and save the timestamp
    5 Thread_2 fails to compare timestamps from steps 2 and 4
    It was fixed in the SimpleResolverProviderImpl, the blocked thread should update the timestamps at the end.


Note that I was not able to reproduce the problems above out of the box, but was able to do that by adding some custom delays here and there in the JDK code.

**Comment by dfuch** [comment]
Thanks - with the latest changes the test appears to be stable.

**Comment by mrserb** [comment]
Does anybody have other comments or the PR is ready to be approved?

**Comment by dfuch** [comment]
Thanks for this later change. I agree with Alan that it does look better.

**Comment by mrserb** [comment]
The CSR is approved, please take a look to the final version.

**Review by dfuch — APPROVED** [review]
LGTM
Having it in 22 early will increase the test exposure.

**Review by Michael-Mc-Mahon — APPROVED** [review]
Looks good.
